src/aspire/classification/rir_class2d.py

```diff
@@ -0,0 +1,577 @@
+        add as a private method and list in nn_implementations.
+
+        :param coef_b:
+        :param coef_b_r:
+        :returns: Array of indices representing classes.
+        """
+        # _nn_classification is assigned during initialization.
+        return self._nn_classification(coef_b, coef_b_r)
+
+    def bispectrum(self, coef):
+        """
+        All bispectrum implementations should consume a stack of fspca coef
+        and return bispectrum coefficients.
+
+        :param coef: complex steerable coefficients (eg. from FSPCABasis).
+        :returns: tuple of arrays (coef_b, coef_b_r)
+        """
+        # _bispectrum is assigned during initialization.
+
+        return self._bispectrum(coef)
+
+    def _sk_nn_classification(self, coeff_b, coeff_b_r, diagnostics=False):
+        # Before we get clever lets just use a generally accepted implementation.
+
+        n_img = self.src.n
+
+        # Third party tools generally expecting:
+        #   slow axis as n_data, fast axis n_features.
+        # Also most third party NN complain about complex...
+        #   so we'll pretend we have 2*n_features of real values.
+        # Don't worry about the copy because NearestNeighbors wants
+        #   C-contiguous anyway... (it would copy internally otherwise)
+        X = np.column_stack((coeff_b.real, coeff_b.imag))
+        # We'll also want to consider the neighbors under reflection.
+        #   These coefficients should be provided by coeff_b_r
+        X_r = np.column_stack((coeff_b_r.real, coeff_b_r.imag))
+
+        # We can compare both non-reflected and reflected representations as one large set by
+        #   taking care later that we store refl=True where indices>=n_img
+        X_both = np.concatenate((X, X_r))
+
+        nbrs = NearestNeighbors(n_neighbors=self.n_nbor, algorithm="auto").fit(X_both)
+        distances, indices = nbrs.kneighbors(X)
+
+        # any refl?
+        logger.info(
+            f"Count reflected: {np.sum(indices>=n_img)}"
+            f" ({np.sum(indices>=n_img) / len(indices)}%)"
+        )
+
+        if diagnostics:
+            # Lets peek at the distribution of distances
+            plt.hist(
+                distances[:, 1:].flatten(), bins="auto"
+            )  # zero index is self, distance 0
+            plt.show()
+
+        # I was considering to change the result of these functions
+        # but for now return classes as range 0 to 2*n_img.
+        # #########################
+        # # There are two sets of vectors each n_img long.
+        # #   The second set represents reflected.
+        # #   When a reflected coef vector is a nearest neighbor,
+        # #   we notate the original image index (indices modulus),
+        # #   and notate we'll need the reflection (refl).
+        # classes = indices % n_img
+        # refl = np.array(indices // n_img, dtype=bool)
+        # corr = distances
+        # return classes, refl, corr
```

Comment:
Because the `legacy_nn_classification` would require some modification to return the same values.  Mainly the distances (`corr`).  Should be harmless to refactor.  I'll give it a go.

Reply:
Turned out fine.  My initial hesitation was both wrt `legacy_nn_classification` and `legacy_align` but the changes were pretty trivial.  Looks better this way to me too.